enhancements/20220822-storage-network/20220822-storage-network.md

```diff
@@ -0,0 +1,163 @@
+And eth2, and eth3 are for VM and Storage Networks.
+
+VM and Storage Network will have different VLAN ID.
+
+#### Story 1
+
+VM and Storage Network will use the same interfaces but different VLAN IDs.
+
+- VM and Storage want to share eth2 and eth3.
+- Create VlanConfig and ClusterNetwork with eth2 and eth3
+- Create Network-Attatchment-Definition for VM
+
+#### Story 2
+
+VM and Storage Network will use different interfaces.
+
+- VM uses eth2 and eth3
+- Storage uses eth4 and eth5
+- Create a new VlanConfig and ClusterNetwork with eth4 and eth5
+
+#### Story 3
+
+Users want to rollback the config back to default
+
+### User Experience In Detail
+
+#### Controller
+
+1. Users have to shutdown all VMs and stop all VMI before setting `storage-network` in Harvester settings
+2. Users need to create a VLAN Config with ClusterNetwork name and ensure the cluster network and VLAN config cover all nodes and network connectivity is expected.
+3. Users have to ensure the Cluster Network is configured and VLAN Config will cover all nodes and ensure the network connectivity is working and expected in all nodes.
+4. Users need to assign ClusterNetwork, VLAN ID, IP Range in storage-network settings
+5. Controller will shut down management pods attached to Longhorn volumes and check all volumes are detached.
+6. Controller will set `storage-network` config to Longhorn settings `storage-network`.
+7. Controller will restart management pods.
+8. Users need to start VM manually.
+
+### API changes
+
+- New Settings `storage-network`
+
+## Design
+
+### Implementation Overview
+
+- Re-use VLAN Bridge from VLAN enhancement.
+- Introduce IPAM [Whereabouts](https://github.com/k8snetworkplumbingwg/whereabouts) to be a global IPAM for Longhorn Pods.
+- Create a Network Attachment Definition with plugin bridge and Whereabouts IPAM.
+- According to [Longhorn storage network limitation](https://longhorn.io/docs/1.3.1/advanced-resources/deploy/storage-network/), users should not modify NAD after setting in Longhorn settings.
+- Controller will create NAD for storage-network, and remove old NAD.
+- Controller will sync storage-network config to Longhorn storage-network setting.
+
+### Controller Flow Chart
+
+![flow chart](./storagenetwork.png)
+
+### Resource needs to be paused or stopped
+
+- Rancher Monitoring
+    - kind: managedchart
+    - namespace: fleet-local
+    - name: rancher-monitoring
+- Grafana
+    - kind: deployments
+    - namespace: cattle-monitoring-system
+    - name: rancher-monitoring-grafana
+- Prometheus
+    - kind: prometheus
+    - namespace: cattle-monitoring-system
+    - name: rancher-monitoring-prometheus 
+- Alertmanager
+    - kind: alertmanager
+    - namespace: cattle-monitoring-system
+    - name: rancher-monitoring-alertmanager
+- Harvester VM Import Controller
+    - kind: deployments
+    - namespace: harvester-system
+    - name: harvester-vm-import-controller
+
+#### Settings config example
+```json
+{
+    "vlan": 100,
```

Comment:
Could the VLAN ID be 0?

Reply:
VLAN should be 1 if users need untagged vlan